File: guides/micronaut-security-x509/java/src/main/java/example/micronaut/HelloController.java
package example.micronaut;

import io.micronaut.core.annotation.Nullable;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.security.annotation.Secured;
import io.micronaut.security.authentication.Authentication;
import io.micronaut.security.x509.X509Authentication;

import static io.micronaut.http.MediaType.TEXT_PLAIN;
import static io.micronaut.security.rules.SecurityRule.IS_ANONYMOUS;

@Controller
class HelloController {

    @Secured(IS_ANONYMOUS) // <1>
    @Get(produces = TEXT_PLAIN) // <2>
    String hello(@Nullable X509Authentication x509Authentication, // <3>
                 @Nullable Authentication authentication) { // <4>
        if (x509Authentication == null && authentication == null) {
            return "Hello unknown!"; // <5>
        }
        if (x509Authentication == null) {
            return "ERROR: Authentication is present but not X509Authentication"; // <6>
        }
        if (x509Authentication != authentication) {
            return "ERROR: Authentication and X509Authentication should be the same instance"; // <7>
        }
        return "Hello " +
                x509Authentication.getName() +
                " (X.509 cert issued by " + x509Authentication.getCertificate().getIssuerX500Principal().getName() + ')'; // <8>
    }
}

File: guides/micronaut-security-x509/python/src/example/micronaut/hello_controller.py
from micronaut.http import MediaType
from micronaut.http.annotation import Get
from micronaut.security.annotation import Secured
from micronaut.security.authentication import Authentication
from micronaut.security.rules import SecurityRule
from micronaut.security.x509 import X509Authentication


@Secured(SecurityRule.IS_ANONYMOUS)  # <1>
@Get(value="/", produces=MediaType.TEXT_PLAIN)  # <2>
def hello(  # <3> <4>
    x509_authentication: X509Authentication = None,
    authentication: Authentication = None,
) -> str:
    if x509_authentication is None and authentication is None:
        return "Hello unknown!"  # <5>
    if x509_authentication is None:
        return "ERROR: Authentication is present but not X509Authentication"  # <6>
    if x509_authentication is not authentication:
        return "ERROR: Authentication and X509Authentication should be the same instance"  # <7>
    issuer = x509_authentication.getCertificate().getIssuerX500Principal().getName()
    return f"Hello {x509_authentication.getName()} (X.509 cert issued by {issuer})"  # <8>
